drivers/xen/regions.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/xen/generic.h>
+
+#include <kernel_arch_interface.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/arch_interface.h>
+#include <zephyr/sys/mem_blocks.h>
+
+LOG_MODULE_REGISTER(xen_regions);
+
+#define EXTENDED_REGIONS_IDX 1
+#define XEN_HYP_NODE DT_INST(0, xen_xen)
+
+static K_MUTEX_DEFINE(ext_regions_lock);
+
+struct extended_regions {
+	sys_mem_blocks_t *block;
+	uint64_t phys_addr;
+	uint64_t size;
+};
+
+/*
+ * At least one extended region should be provided.
+ * If there is no extended regions - alternative allocation should be implemented
+ * by the driver user and driver should be disabled.
+ */
+BUILD_ASSERT(DT_NUM_REGS(XEN_HYP_NODE) >= EXTENDED_REGIONS_IDX,
+	     "No Extended regions were provided in DT");
+
+/*
+ * Add BUILD_ASSERT that rejects extended region, starting from 0x0.
+ * This is done because mem_blocks is not designed to handle buffer
+ * starting from 0 and has assert with check. In this case starting address
+ * should be shifted to page 1. That shouldn't be a problem because
+ * minimal extended region size is set to 64MB in XEN.
+ * For example for the extended region 0->0x30000000 the following
+ * regs should be provided:
+ * regs = <.... 0x0 0x1000 0x0 0x2FFFF000 ... >;
+ *
+ * Check reg addr and reg size to be aligned to
+ * CONFIG_MMU_PAGE_SIZE. This is the requirement for arch_mem_map call.
+ *
+ * Extended regions should not overlap kernel Virtual Memory region.
+ */
+#define EXT_BLOCK(node_id, idx)						\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),		\
+		     "Extended region starting from 0 is not supported, please skip page 0"); \
+	BUILD_ASSERT(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region size is not aligned");		\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region address is not aligned");		\
+	BUILD_ASSERT((DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) < CONFIG_KERNEL_VM_BASE || \
+		      DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) >=		\
+		      CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE),	\
+		     "Extended region should not overlap virtual ram");	\
+	BUILD_ASSERT(((DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) +		\
```

Comment:
(major) incorrect checks inside build asserts, lets imagine the next situation:
DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) == 0x1000
DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)) == 0x1000
CONFIG_KERNEL_VM_BASE == 0x1100
CONFIG_KERNEL_VM_SIZE == 0x100

first assert:
BUILD_ASSERT(0x1000 < 0x1100 || 0x1000 >= 0x1100 + 0x100)
here we don't get assert because first condition will be true, e.g. `0x1000 < 0x1100`

second assert:
BUILD_ASSERT(0x1000 + 0x1000 < 0x1100 || 0x1000 + 0x1000 >= 0x1100 + 0x100)
here first condition will be **false**, but second will be **true** --- 0x2000 > 0x1200

so VM region will be inside you extended region, i.e. there is a full overlap

Reply:
just google how to find that two ranges have intersections there is a lot of good solutions with a small amount of checks

Reply:
What made you think that I didn't googled before saying about the checks?

Reply:
rewritten